### PR TITLE
Fix #761, #762: support nullable identifier hint in proxies

### LINKED ISSUE #761: ProxyGenerator::isShortIdentifierGetter in PHP 7.1 does not support nullable types
ProxyGenerator::isShortIdentifierGetter does not work with PHP7.1 nullable return types.

Does not work
```
public function getId() : ?int
```

Works

```
public function getId() : int
```

If I understood correctly the `ProxyGenerator::PATTERN_MATCH_ID_METHOD` should be updated to include question mark in return type and test cases to have case with nullable return type.

### PR DESCRIPTION
**Pull request by Ocramius**
Fixes #761
Fixes #762
Replaces #762 (includes its commits)

### CODE REVIEW

File: tests/Doctrine/Tests/Common/Proxy/LazyLoadableObjectWithNullableTypehints.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+    /** @var \stdClass */
+    private $identifierFieldReturnClassOneLetterNullableWithSpace;
+
+    public function getIdentifierFieldReturnClassOneLetterNullable(): ?A
+    {
+        return $this->identifierFieldReturnClassOneLetterNullable;
+    }
+
+    public function getIdentifierFieldReturnClassOneLetterNullableWithSpace(): ? A
```
Comment (Soullivaneuh) [inline tests/Doctrine/Tests/Common/Proxy/LazyLoadableObjectWithNullableTypehints.php]: Don't know what is the correct syntax, but you set `? A` here and `?A`.

I suppose this is not wanted. :wink:

Comment (Ocramius) [inline tests/Doctrine/Tests/Common/Proxy/LazyLoadableObjectWithNullableTypehints.php]: Space is intentional, since both approaches are valid/accepted

Comment (Soullivaneuh) [inline tests/Doctrine/Tests/Common/Proxy/LazyLoadableObjectWithNullableTypehints.php]: Oh indeed.

But why do you have to test both syntaxes? The behavior is same, am I wrong?

Comment (Ocramius) [inline tests/Doctrine/Tests/Common/Proxy/LazyLoadableObjectWithNullableTypehints.php]: Yes, but the code analyzer is regex-based (yeah, I know :S)

Comment (Soullivaneuh) [inline tests/Doctrine/Tests/Common/Proxy/LazyLoadableObjectWithNullableTypehints.php]: > Yes, but the code analyzer is regex-based

Damn me, indeed!

> (yeah, I know :S)

Regex is good, eat it! :trollface:

**Comment by Ocramius** [comment]
Backported into `2.7` in doctrine/common@930297026c8009a567ac051fd545bf6124150347

**Pull request merged**